tiatoolbox/dataloader/wsireader.py

```diff
@@ -82,39 +91,524 @@ def slide_info(self):
+            start_w (int): starting point in x-direction (along width).
+            start_h (int): starting point in y-direction (along height).
+            end_w (int): end point in x-direction (along width).
+            end_h (int): end point in y-direction (along height).
+            resolution (int or float or tuple of float): resolution at
+                which to read the image, default = 0. Either a single
+                number or a sequence of two numbers for x and y are
+                valid. This value is in terms of the corresponding
+                units. For example: resolution=0.5 and units="mpp" will
+                read the slide at 0.5 microns per-pixel, and
+                resolution=3, units="level" will read at level at
+                pyramid level / resolution layer 3.
+            units (str): the units of resolution, default = "level".
+                Supported units are: microns per pixel (mpp), objective
+                power (power), pyramid / resolution level (level),
+                pixels per baseline pixel (baseline).
 
         Returns:
-            ndarray : image array
+            ndarray : array of size MxNx3
+            M=end_h-start_h, N=end_w-start_w
 
+        Examples:
+            >>> from tiatoolbox.dataloader import wsireader
+            >>> from matplotlib import pyplot as plt
+            >>> wsi_obj = wsireader.WSIReader("/path/to/a/wsi")
+            >>> # Read a region at level 0 (baseline / full resolution)
+            >>> bounds = [1000, 2000, 2000, 3000]
+            >>> img = wsi_obj.read_bounds(*bounds)
+            >>> plt.imshow(img)
+            >>> # This could also be written more verbosely as follows
+            >>> img = wsi_obj.read_bounds(
+            ...     start_w=bounds[0],
+            ...     start_h=bounds[1],
+            ...     end_w=bounds[2],
+            ...     end_h=bounds[3],
+            ...     resolution=0,
+            ...     units="level",
+            ... )
+            >>> plt.imshow(img)
+
+        Note: The field of view remains the same as resolution is varied.
+
+        .. figure:: images/read_bounds_tissue.png
+            :width: 512
+            :alt: Diagram illustrating read_bounds
+
+        This is because
+        the bounds are in the baseline (level 0) reference
+        frame. Therefore, varying the resolution does not change what is
+        visible within the output image.
+
+        If the WSI does not have a resolution layer
+        corresponding exactly to the requested resolution
+        (shown above in white with a dashed outline), a larger
+        resolution is downscaled to achieve the correct requested output
+        resolution.
+
+        If the requested resolution is higher than the
+        baseline (maximum resultion of the image), then bicubic
+        interpolation is applied to the output image.
         """
-        raise NotImplementedError
+        raise NotImplementedError()
```

Comment:
what is the difference between `NotImplementedError` and `NotImplementedError()`

Reply:
They are equivalent. I believe these are all equivalent:

```python
raise NotImplementedError
raise NotImplemenetedError()
raise(NotImplementedError)
raise(NotImplemenetedError())
```

I don't remember changing this. Happy to decide on what to do when there is no message argument to an exception.

Reply:
Let's use raise NotImplementedError